### ISSUE TITLE
deploy apps without ELBs (but with ports exposed) - for external service discovery

### ISSUE THREAD

**Issue by soulware** [convox/rack#492]
Thinking about service discovery in the context of internal (micro)-services.

We _can_ use the convox created load balancers (one per service) but at some point this might become unwieldy (_lots_ of ELBs).
There is also no real service discovery going on - you still need to know either the endpoints
of the load balancers or set DNS up to map cleanly to them.

I was wondering if there is a more lightweight/simpler approach where convox apps could (optionally) be deployed with ports exposed but without the load balancers - and leave service discovery up to an external setup like Consul+Registrator or something similar.

https://www.consul.io/
https://github.com/gliderlabs/registrator

It would be great if we could -
- deploy a service with port(s) exposed, but without an ELB being created
- let ECS choose a random port (so we could deploy more than one task per EC2 instance)

From the ECS docs -

https://aws.amazon.com/blogs/compute/service-discovery-via-consul-with-amazon-ecs/

```
"portMappings": [
  {
    "containerPort": 4567
  }
]
```

> Notice that you are not defining the host port mapping. Docker automatically assigns a port on the host so the port number is discovered via the Consul service discovery. This allows you to run multiple tasks or services of the same type on a single ECS instance.

There is some mention of host port "0" here -
https://forums.aws.amazon.com/message.jspa?messageID=596662

> If you leave out the host port, or specify it as being '0', in the task definition, then the port will be chosen automatically and reported back to the service.  After the task is running, describing it should show what port(s) it bound to.

So maybe one way to do this via `docker-compose.yml` would be to allow
ports to be specified something like -

```
# public ELB mapping port 80 -> 80
ports:
  - 80:80

# internal ELB mapping port 80 -> 80
ports:
  - 80

# *no* ELB and random host port in ECS
ports:
  - 0:80
```

Thoughts?

**Comment by ShakataGaNai** [comment]
I'm a fan of NOT creating an ELB for every service. For one or two apps that's ok, but if running say 10 different "light" projects, you'd end up paying $200/mo just for ELB's.

Perhaps another option is one ELB pointing at an internal load-balancer that's then integrated into the various applications?

**Comment by Crashthatch** [comment]
+1 for this. I just started using Convox, but I'm mostly using it to host several of my own weekend projects. A project I hacked together in a weekend that just needs to run somewhere with a port or 2 exposed to the internet isn't worth $18 / month (tbh, for me each project is probably not worth more than $2 / month or after a year of weekends I'd be paying hundreds per month).

**Comment by nzoschke** [comment]
For service discovery, this recent post caught my attention: https://aws.amazon.com/blogs/compute/service-discovery-an-amazon-ecs-reference-architecture/

This particularly jumps out as the Convox way:

> service registration is extremely affordable, instantaneous, reliable, and maintenance-free.

Over needing to set up and operate Consul. 

@soulware exposing a port without creating an ELB does make sense though, and the port 0 design seems good on first pass.

Another design would be to use a label like we do for other load balancer settings: https://convox.com/docs/load-balancers/

@ShakataGaNai @Crashthatch The cost of ELBs bugs me too, but at the moment Convox simply is focusing on getting serious production apps set up right. It simply isn't designed for hobby apps.

There is no plan to build an internal load balancing mechanism into Convox right now. The justification is that it's tricky to build a system that will keep everything in sync and meet the throughput and availability properties we expect from ELBs.

Finally, this is speculative, but I suspect AWS will add single ELB and ECS integration in the future or release a smaller/cheaper container ELB service. It's such a clear wart in the system, just like the previous lack of a Registry and CloudWatch Logs, that they almost certainly will improve.

**Comment by altaurog** [comment]
This is all very interesting.  IMO an ELB-based solution has some severe limitations.

1. Exposing services meant purely for internal use via ELB and Route 53 just for service discovery may be undesirable. (Is it possible to avoid external exposure with this approach?)
2. This solution doesn't address integration of servers which aren't HTTP servers.

The architecture I am considering for a project currently on our drawing board has clients which receive events from an external source, an event queue, and workers which process the events.  I don't imagine any of the three components will be serving HTTP.  At a minimum, both the event clients and the workers will need to know how to reach the event queue.  We could use ElastiCache redis for queuing and bake the address in, but I am not certain that redis is the right tool for the job.  We may want to run some other distributed queue within the ECS cluster, in which case also the queue nodes will need service discovery to be able to participate.

Another issue ELB wouldn't solve is that event clients also need to be assigned "topics."  When a client fails and is replaced by a new client, the new client needs to be handed the correct topic.  (I think of that as a service discovery problem but realize it may be out of scope for this issue.  In any case, I am definitely open to suggestions!)

**Comment by mwarkentin** [comment]
@altaurog You can create an "internal" service by defining a single port in your docker compose file:

```
ports:
  - 80
```

If you want to expose it externally, it's defined like this:

```
ports:
  - 80:80
```

In the former case, it will create an "internal" ELB which is only accessible from within the VPC - but can still handle the load balancing / "discovery" for balancing across multiple processes.

By default, the ELB listener is actually a TCP listener, not HTTP - not sure if that answers your second question.

**Comment by altaurog** [comment]
@mwarkentin thanks, I missed that.

**Comment by nzoschke** [comment]
@altaurog this is off topic but have you considered SQS for your architecture? `convox services create sqs` makes it easy to experiement with, and totally avoids service discovery and operating your own queue :)

**Comment by altaurog** [comment]
Thanks, Noah.  Yes I have considered both SQS and Kinesis.  There are other considerations and I'm more concerned at the moment with how to manage the configuration of the listeners and the workers.  Perhaps I could follow up with you next week about that?

**Comment by nzoschke** [comment]
@altaurog yeah, feel free to reach out directly on slack or email. With `convox services create sqs` you get an SQS_URL that you copy/paste into your apps. Since its a managed service the URL is effectively static avoiding service discovery things.

Convox could use a little more sugar to enable `convox service link mysqs --app myapp` for some of the data store services.

https://github.com/convox/rack/issues/466

**Comment by spencerhakim** [comment]
This issue can probably be closed now that #1535 has been merged and released.

**Comment by nzoschke** [comment]
Yes, this is no possible with #1535. Thanks for the report and the patch.
